### ISSUE TITLE
buffer operator or toArray

### ISSUE THREAD

**Issue by jvanbruegge** [staltz/xstream#142]
It would be very helpful to be able to accumulate all values held back by debounce in an array (like in RxJS). That way, you can implement triple clicks or similar behavior far easier.

**Comment by xtianjohns** [comment]
Hey @SuperManitu, I'm not sure of your exact use case with debounce, but we should be able to implement aggregation and buffering by composing several xstream core and extra operators.

For example, if you needed debounce to "buffer" values and emit them as a group (like RxJS `buffer`), you can use xstream's [`split`](https://github.com/staltz/xstream/blob/master/EXTRA_DOCS.md#split) extra. It behaves very similarly to Rx's [`window`](https://github.com/Reactive-Extensions/RxJS/blob/master/doc/api/core/operators/window.md) operator.

Here is an example of using it to split click events into counts that occur within a time window:

```ts
import xs from 'xstream';
import flattenSequentially from 'xstream/extra/flattenSequentially';
import fromEvent from 'xstream/extra/fromEvent';
import split from 'xstream/extra/split';

/* Get the button. */
const button = document.getElementById( 'button' );

/* Create a .5 second timer. */
const halfSecond = xs.periodic( 500 );

/* Capture all click events on a button. */
const allClicks = fromEvent( button, 'click' );

// Group events in the click stream.
const counts = allClicks
  .compose( split( halfSecond ) )
  .map( stream => stream.fold( count => count + 1, 0 ).last() )
  .compose( flattenSequentially );
```
Then, for a triple click implementation (which you mentioned), you could filter these counts.
```ts
const triples = counts.filter( clicks => clicks > 2 ); // triple click happened
```
Now, arguably, this doesn't get you to a true "triple-click" because you could click three times during an overlap between two half-second periods. But an alternative approach can fix that:
```ts
// Emit if 3 clicks happen in .5 seconds.
const triples = buffer( allClicks, 3, halfSecond );

function buffer( source, count, closing ) {
  const grouped = source.map( item => {
    // For every click, track the next two clicks.
    const buffered = source.take( count - 1 );
    // End the stream if 'closing' emits first.
    return buffered.endWhen( closing ).last();
  });
  return grouped.compose( flattenSequentially );
}
```
So, depending on one's needs, composing operators can create some really powerful functionality.

If either of these fit the use cases you had in mind, do you still think it's beneficial to implement a change to `debounce` or add an extra `buffer` operator?

@staltz, do you think there's value in an operator that emulates `buffer`, or `buffer`-esque operators from other libraries? In the first case, functionality was achievable with two extra operators and four extra lines of code. In the second, it was two operators and seven lines.

Are there use cases in Cycle.js that would benefit from a shorter/more ergonomic implementation?

**Comment by staltz** [comment]
Thanks @xtianjohns for the comprehensive comment.

`debounce` won't get any options, but we should consider making a `toArray` **extra** operator. It would accumulate all values in a stream in an array, and then emit that array when the source completes.

```
--a----b----c--d----------|
      toArray
--------------------------[a,b,c,d]|
```

This would make it easy to build `buffer` alternatives. Typical buffer would be `source.split(y$).map(toArray).flatten()`.

We could bikeshed around the name `toArray`. I'm annoyed that it may seem that it returns `Array<T>`, when in fact it returns `Stream<Array<T>>`. I thought of other names like `collect`, `toArrayStream`, `collectAsArray`, etc, but I'm not sure if they make it more understandable than `toArray` which already exists in RxJS.

**Comment by SuperManitu** [comment]
I like the toArray aproach, as it also allows to use streams as better arrays when dealing with a lot of data and a lot of transformation (no itermediary arrays like array.map). And together with split exacly the needed behavior.
One thing I dont understand is ```flatten()``` in your example. Shouldnt it be like this?:
```
source //Stream<T>
    .compose(split(y$)) //Stream<Stream<T>>
    .compose(toArray) //Stream<T[]>
```

**Comment by staltz** [comment]
@SuperManitu Hi. No because the type signature for toArray must be Stream<T> to Stream<Array<T>> because we can't synchronously collect all (possibly async) values of a stream into an array.

**Comment by xtianjohns** [comment]
Should this be closed by #196, or do we still want `toArray`?

**Comment by staltz** [comment]
Yes this is closed by #196 and `toArray` can be emulated with `.compose(buffer(xs.never()))`.
